security/oidc: Add metrics, JWKS refresh and Some admin endpoints

BenPope opened the pull request:
## Documentation:
### New cluster configuration options:
| option | description | default |
| -- | -- | -- |
| `oidc_keys_refresh_interval` | "The frequency of refreshing the JSON Web Keys (JWKS) used to validate access tokens." | `1h` |

### New Admin API endpoins:
```json
"/v1/security/oidc/whoami": {
    "get": {
        "summary": "Obtain the principal and details of the JWT passed in the bearer token",
        "operationId": "oidc_whoami",
        "responses": {
            "200": {
                "description": "Successful response",
                "schema": {
                    "description": "Details of the provided bearer token.",
                    "type": "object",
                    "properties": {
                        "id": {
                            "description": "Mapped principal for autherization.",
                            "type": "string"
                        },
                        "expire": {
                            "description": "Unix timestamp of the expiry of the token.",
                            "format": "uint64",
                            "type": "int"
                        }
                    }
                }
            },
            "401": {
                "description": "Unauthorized"
            }
        }
    }
},
"/v1/security/oidc/keys/cache_invalidate": {
    "post": {
        "summary": "Reload the keys from the Identity Provider",
        "operationId": "oidc_keys_cache_invalidate",
        "responses": {
            "200": {
                "description": "Successful response"
            },
            "500": {
                "description": "Keys failed to reload",
                "schema": {
                     "type": "object",
                     "properties": {
                         "error_message": {
                             "description": "The error message.",
                            "type": "string"
                        } 
                    }
                }
            }
        }
    }
}
```

### New metrics

E.g.:
```
# HELP redpanda_security_idp_errors_total Number of requests that returned with an error.
# TYPE redpanda_security_idp_errors_total counter
redpanda_security_idp_errors_total{redpanda_domain="dev-ltxchcls4igzho78.us.auth0.com",redpanda_mechanism="oidc",shard="0"} 0
redpanda_security_idp_errors_total{redpanda_domain="dev-ltxchcls4igzho78.us.auth0.com",redpanda_mechanism="oidc",shard="1"} 0
# HELP redpanda_security_idp_latency_seconds Latency of requests to the Identity Provider
# TYPE redpanda_security_idp_latency_seconds histogram
redpanda_security_idp_latency_seconds_sum{redpanda_domain="dev-ltxchcls4igzho78.us.auth0.com",redpanda_mechanism="oidc",shard="0"} 0.672952
redpanda_security_idp_latency_seconds_count{redpanda_domain="dev-ltxchcls4igzho78.us.auth0.com",redpanda_mechanism="oidc",shard="0"} 2
redpanda_security_idp_latency_seconds_bucket{le="0.000255",redpanda_domain="dev-ltxchcls4igzho78.us.auth0.com",redpanda_mechanism="oidc",shard="0"} 0
```

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v23.2.x
- [ ] v23.1.x
- [ ] v22.3.x

## Release Notes

* none

vbotbuildovich commented:
<!--NEW FAILURES-->
new failures in https://buildkite.com/redpanda/redpanda/builds/41193#018bd5d1-c8a2-4207-a2d4-436cfaf58ece:
```
"rptest.tests.audit_log_test.AuditLogTestOauth.test_kafka_oauth"
```
<!--f2e57d03-2b25-4a89-bc52-dae6ea8daac2-->

oleiman commented on src/v/redpanda/admin/server.cc:
```diff
@@ -2106,6 +2108,31 @@ admin_server::update_user_handler(std::unique_ptr<ss::http::request> req) {
+    }
+
+    security::oidc::authenticator auth{_controller->get_oidc_service().local()};
+    auto res = auth.authenticate(auth_hdr.substr(7));
```
>auth_hdr.substr(7)

nit: probably overkill to add any real abstraction, but the number `7` here contains arcane knowledge. more an observation than a nit I guess. (it's just len("Bearer") right?)

BenPope replied:
> > auth_hdr.substr(7)
> 
> nit: probably overkill to add any real abstraction, but the number `7` here contains arcane knowledge. more an observation than a nit I guess. (it's just len("Bearer") right?)

Yeah, I've been wondering about this a bit, but it seemed overkill. I'll add the abstraction.

BenPope replied:
>> the number 7 here contains arcane knowledge
>
> I don't think it's really worth writing the abstraction.

Added some constants to request_auth and used them.

oleiman reviewed (commented):
looks good to me! just a few questions

BenPope commented on src/v/security/oidc_service.cc:
```diff
@@ -72,84 +76,85 @@ struct service::impl {
         _jwks.emplace(std::move(jwks).assume_value());
+
+        auto res = _verifier.update_keys(*_jwks);
+        if (res.has_error()) {
```
> Would it make sense to clear out the parsed but invalid JWKS?

Good spot, I'll remove the `_jwks` member.

BenPope replied:
Done

BenPope commented on src/v/security/oidc_service.cc:
```diff
@@ -72,84 +76,85 @@ struct service::impl {
-        if (res.has_error()) {
+    ss::future<errc> update_metadata() {
+        auto response_body = co_await ss::coroutine::as_future(
+          make_request(*_parsed_discovery_url));
```
> since its assignment and use are not colocated, would it make sense to assert on _parsed_discovery_url containing a value?

I'll colocate them.

> Similar comment for the probe being non-null.

I'm not sure which bit you're referring to.

oleiman replied:
This line (sorry I couldn't figure out how to get a permalink from the core review window): https://github.com/redpanda-data/redpanda/pull/14990/files#r1395970819

BenPope commented on src/v/security/oidc_service.cc:
```diff
@@ -72,84 +76,85 @@ struct service::impl {
-            vlog(seclog.error, "Error updating keys: Keys not found");
-            co_return;
-        }
+        co_await update_metadata();
+        co_await update_jwks();
```
I'll refactor the error handling. `future<errc>` isn't very nice.

Assume there is a `jwks_uri` (which is checked), a refresh is still acceptable; the `jwks_uri` is unlikely to change, which is the only relevant part of `update_metadata`.

BenPope commented on src/v/security/oidc_service.cc:
```diff
@@ -22,9 +24,84 @@
+        ++_request_errors;
+    }
+
+    // record the measurement
+    void request_success(hist_t::measurement) {}
```
I'll improve the comment.

BenPope replied:
Done

oleiman commented on src/v/security/oidc_service.cc:
```diff
@@ -129,84 +181,106 @@ struct service::impl {
-        auto res = _verifier.update_keys(*_jwks);
-        if (res.has_error()) {
+    ss::future<errc> update_metadata() {
+        auto& probe = _probes[_parsed_discovery_url->host];
```
Guaranteed to be non-null?

BenPope replied:
> Guaranteed to be non-null?

Yes, but added `get_probe_for` anyway.

michael-redpanda commented on src/v/security/exceptions.h:
```diff
@@ -22,9 +22,8 @@ namespace security {
-    explicit exception(errc e, std::string_view msg)
-      : std::runtime_error(
-        fmt::format("{}: {}", make_error_code(e).message(), msg)) {}
+    explicit exception(std::error_code e, std::string_view msg)
```
question: why do youuse `std::runtime_error` as a base rather than `std::system_error` which accepts `std::error_code`'s.  I've noticed this pattern and was just curious ;)

BenPope replied:
> question: why do youuse `std::runtime_error` as a base rather than `std::system_error` which accepts `std::error_code`'s. I've noticed this pattern and was just curious ;)

Just making minimal changes to `security::exception` which has been around for a while, I could make a new exception type in `security::oidc`.

michael-redpanda replied:
up to you, was just curious :)

BenPope replied:
Kept as-is

BenPope commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/ae4ca20a52623006259161087a796f27fa5523e1..49dd4f4931186687542e0d1275f5a0fd784ce57a)

* Added constants for authz header
* Improved error handling in `oidc_service::update` - used `security::exception` instead of `future<errc>`, and then deferred the logging.
* Other smaller review comments

oleiman commented on src/v/security/oidc_service.cc:
```diff
@@ -19,67 +23,100 @@
-    if (url_res.has_error()) {
-        throw std::runtime_error("Invalid url");
+
+template<typename... Args>
+[[nodiscard]] auto return_exception(
+  std::error_code ec,
+  fmt::format_string<Args...> fmt,
+  Args&&... args) noexcept {
+    return ss::coroutine::return_exception(
+      exception(ec, fmt::format(fmt, std::forward<Args>(args)...)));
+}
```
neat

oleiman reviewed (approved):
lgtm!

michael-redpanda commented on src/v/redpanda/admin/server.cc:
```diff
@@ -2106,6 +2108,31 @@ admin_server::update_user_handler(std::unique_ptr<ss::http::request> req) {
+    security::oidc::authenticator auth{_controller->get_oidc_service().local()};
+    auto res = auth.authenticate(auth_hdr.substr(authz_bearer_prefix.length()));
+
+    if (res.has_error()) {
+        throw ss::httpd::base_exception{
+          "Invalid Authorization header",
+          ss::http::reply::status_type::unauthorized};
+    }
```
question: would it be any good to include the reason why the token was invalid in the response (I think `res` may contain such info)

BenPope replied:
> question: would it be any good to include the reason why the token was invalid in the response (I think `res` may contain such info)

Since the endpoint is `whoami`, and the `authentication_header` has been taken from the request, it would already have performed a similar check and returned prior to getting here. Even if it did get here, I don't think it's wise to expose too much detail as to why a request failed, it gives an attacker information to reduce their search space.

If the request was more like `finger` then it would make sense for a superuser to be able to figure out what's wrong with the token, after they have passed their own auth. That was ruled out as a requirement.

michael-redpanda reviewed (approved):
looks good to me

BenPope commented on src/v/security/oidc_service.cc:
```diff
@@ -41,6 +44,76 @@ template<typename... Args>
+class probe {
+public:
+    using hist_t = log_hist_public;
+
```
I didn't get the build failure locally, but the fix is something like:
```c++
    class auto_measure {
    public:
        explicit auto_measure(probe* p)
          : _p{p}
          , _m{p->_request_latency} {}
    public:
        void success() noexcept { _p->request_success(std::move(_m)); }
        void failed() noexcep`t { _p->request_error(std::move(_m)); }
    private:
        probe* _p;
        hist_t::measurement _m;
    };
```
And then used like:
```diff
-            probe->request_error(std::move(measure));
+            measure.failed();
```

oleiman replied:
Change looks good, but I'm curious to know what the failure looked like - I don't see it anywhere.

BenPope replied:
For example:
```c++
        if (response_body.failed()) {
            probe->report_error(std::move(measure));
            co_await return_exception(               // Note, this is really a return
              errc::metadata_invalid,
              "Failed to retrieve metadata: {}, error: {}",
              _discovery_url(),
              response_body.get_exception());
        }
        auto metadata = oidc::metadata::make(response_body.get());
        if (metadata.has_error()) {
            probe->report_error(std::move(measure));  // Use after move 
            co_await return_exception(
              metadata.assume_error(),
              "invalid response from discovery_url: {}, err: {}",
              _discovery_url(),
              metadata.assume_error().message());
        }
```

oleiman replied:
Oh yeah, I understand the issue. I thought you meant it was flagged in CI somehow but not locally. Nbd.

BenPope replied:
> I thought you meant it was flagged in CI somehow but not locally

That was the case: https://buildkite.com/redpanda/redpanda/builds/41244#018bd917-04ca-425d-85af-e02710a75b54/474-1023

BenPope commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/49dd4f4931186687542e0d1275f5a0fd784ce57a..58828b1178837265d11f3e130f3d506beb342d7b)
* Fix `use-after-move`.

oleiman reviewed (approved):
nice

michael-redpanda commented on src/v/security/oidc_service.cc:
```diff
@@ -19,67 +23,113 @@
+public:
+    using hist_t = log_hist_public;
+
+    class auto_measure {
+    public:
+        explicit auto_measure(probe* p)
+          : _p{p}
+          , _m{p->_request_latency} {}
+
+    public:
+        void success() noexcept { _p->request_success(std::move(_m)); }
+        void failed() noexcept { _p->request_error(std::move(_m)); }
+
+    private:
+        probe* _p;
+        hist_t::measurement _m;
+    };
```
Not trying to be too pedantic here, but should you have some sort of protection to ensure `success()` and `failure()` are not called multiple times? 
```suggestion
    class auto_measure {
    public:
        explicit auto_measure(probe* p)
          : _p{p}
          , _m{p->_request_latency} {}

    public:
        void success() noexcept {
          if (!_reported) {
            _p->request_success(std::move(_m)); 
            _reported = true;
          }
        void failed() noexcept { 
          if (!_reported) {
            _p->request_error(std::move(_m)); 
            _reported = true;
          }
        }

    private:
        bool _reported{false};
        probe* _p;
        hist_t::measurement _m;
    };
```

michael-redpanda replied:
Or have the user set a flag and do the remove in the dtor?

BenPope replied:
It's a fair point..

BenPope replied:
Re-engineered `probe`.

BenPope commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/58828b1178837265d11f3e130f3d506beb342d7b..400642fd4affa884fd0eed5acdcd1e8a50dce6b5):
* `probe`
  *  `auto_measure` is now harder to misuse
  * Removed `request_error` and `request_success`
* `whoami`
  * Improved ducktape testing
* `cache_invalidate`
  * Invalidate cache on all cores
  * Added ducktape testing

BenPope commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/400642fd4affa884fd0eed5acdcd1e8a50dce6b5..5080140ebfd24713a3c1df104979fdeb79cfc5dd):
* Fix the new `whoami` tests

BenPope commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/5080140ebfd24713a3c1df104979fdeb79cfc5dd..6757950226cbf9947689d773a5ffa27b5849320c):
* Lint the new `whoami` tests!

BenPope commented:
Changes in `force-push`:
* Now that the credentials outlive the method, only supply the credentials if the request is https.

BenPope commented:
This "fixes" the tests that failed:

"rptest.tests.audit_log_test.AuditLogTestOauth.test_kafka_oauth"
"rptest.tests.audit_log_test.AuditLogTestOauth.test_admin_oauth"

But those tests are broken; they don't set these config params:
```
            extra_rp_conf={
                "oidc_discovery_url": self.keycloak.get_discovery_url(kc_node),
                "oidc_token_audience": TOKEN_AUDIENCE
            },
```

michael-redpanda commented:
> This "fixes" the tests that failed:
> 
> `rptest.tests.audit_log_test.AuditLogTestOauth.test_kafka_oauth` `rptest.tests.audit_log_test.AuditLogTestOauth.test_admin_oauth`
> 
> But those tests are broken; they don't set these config params:
> 
> ```
>             extra_rp_conf={
>                 "oidc_discovery_url": self.keycloak.get_discovery_url(kc_node),
>                 "oidc_token_audience": TOKEN_AUDIENCE
>             },
> ```

Sorry about that (I wrote those).  Want me to fix them or are you going to in this PR?

BenPope commented:
> > This "fixes" the tests that failed:
> > 
> > `rptest.tests.audit_log_test.AuditLogTestOauth.test_kafka_oauth` `rptest.tests.audit_log_test.AuditLogTestOauth.test_admin_oauth`
> > 
> > But those tests are broken; they don't set these config params:
> > 
> > ```
> >             extra_rp_conf={
> >                 "oidc_discovery_url": self.keycloak.get_discovery_url(kc_node),
> >                 "oidc_token_audience": TOKEN_AUDIENCE
> >             },
> > ```
> 
> Sorry about that (I wrote those).  Want me to fix them or are you going to in this PR?

I'll investigate now, for a follow-up PR

oleiman commented on src/v/security/oidc_service.cc:
```diff
@@ -19,67 +23,114 @@
+    public:
+        void success() noexcept {
+            if (std::exchange(_p, nullptr) != nullptr) {
+                auto report_and_destroy = std::move(_m);
```
What's to prevent this move assignment from getting blown away by an optimization pass?

BenPope replied:
> What's to prevent this move assignment from getting blown away by an optimization pass?

Whilst copy and move constructors are allowed to be elided in some cases, this is not one of them. The destructor must be called at the end of this scope, or am I missing something?

oleiman replied:
No, that sounds right. Thanks

BenPope commented on tests/rptest/tests/redpanda_oauth_test.py:
```diff
@@ -248,6 +248,49 @@ def test_java_client(self):
+            )
+            return response
+
+        # At this point, admin API does not require auth and service_user_id is not a superuser
```
> nit: I'm having a hard time connecting this comment with subsequent operations. Regardless of cluster config, don't you need a bearer token for a well-formed `whoami` request?

The endpoint requires `user` permission (c.f., `superuser`, `publik`), so I thought it worth noting whether Auth was enabled (every request is allowed if auth is off off), and to note that `service_user_id` is not a superuser (to distinguish it distinguish `superuser` and `user` credentials).

When Admin API authn is enabled, the `unauthorised` status code  is returned by the authn code that wraps the handler
When Admin API authn is disabled, the `unauthorized` status code is returned by the mostly duplicated code in the auth handler.

BenPope commented on src/v/redpanda/admin/server.cc:
```diff
@@ -2133,6 +2135,20 @@ admin_server::oidc_whoami_handler(std::unique_ptr<ss::http::request> req) {
+    auto f = co_await ss::coroutine::as_future(
+      _controller->get_oidc_service().invoke_on_all(
+        [](auto& s) { return s.refresh_keys(); }));
+    if (f.failed()) {
+        ss::httpd::security_json::oidc_keys_cache_invalidate_error_response res;
+        res.error_message = ssx::sformat("", f.get_exception());
+        co_return ss::json::json_return_type(res);
+    }
```
> I guess the only action available to a client is to retry, so an intermediate state (refresh failed, some but not all key stores updated) shouldn't have any ill effect, right?

That was my reasoning. An optimisation is to do all of the HTTP requests to the IdP on a single core and then invoke on all impls, the application of the result of the HTTP request,

oleiman replied:
Makes sense to me

oleiman reviewed (approved):
lgtm!

BenPope commented:
> > > This "fixes" the tests that failed:
> > > `rptest.tests.audit_log_test.AuditLogTestOauth.test_kafka_oauth` `rptest.tests.audit_log_test.AuditLogTestOauth.test_admin_oauth`
> > > But those tests are broken; they don't set these config params:
> > > ```
> > >             extra_rp_conf={
> > >                 "oidc_discovery_url": self.keycloak.get_discovery_url(kc_node),
> > >                 "oidc_token_audience": TOKEN_AUDIENCE
> > >             },
> > > ```
> > 
> > 
> > Sorry about that (I wrote those).  Want me to fix them or are you going to in this PR?
> 
> I'll investigate now, for a follow-up PR

@michael-redpanda - it is set, just not where I expected to see it. The config is modified, which explains why there were failures connecting to the default provider at startup (which is `https`), and then the GnuTLS errors later on (`_creds` were unconditionally being passed, Keycloak isn't set up with TLS yet).

https://github.com/redpanda-data/redpanda/commit/ddd54ced6ea5a24461727212296523ccae26d47c#diff-8cd61a647d6a3bfd57a7c6cbad1916cb495d986c91adfcbbf5bd8130d7107871R1117-R1122

So nothing to worry about!

oleiman commented:
>Keycloak isn't set up with TLS yet

Right...meant to make a ticket for that earlier this week. Here's one: #15020

michael-redpanda reviewed (approved):
lgtm
